File: basics/01-array/0112-704.py
"""
leetcode 704 in python
"""

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums)-1
        # [left, right]
        # both sides are inclusive!
        while left <= right: # <= is a must
            mid = left + (right-left)//2 ## // is correct way
            if nums[mid] > target:
                # mid is not the answer
                right = mid-1
            elif nums[mid] < target:
                # mid is not the answer for sure
                left = mid+1
            else:
                return mid
        return -1

File: basics/01-array/0112-704.go
/*
 * leetcode 704 in Go
 * recursion
 *
 */

func search(nums []int, target int) int {
    return bSearch(nums, target, 0, len(nums)-1);    
}

func bSearch(nums []int, target int, left int, right int) int {
    // base case to stop
    if (left > right) {
        return -1
    }
    mid := left + (right-left)/2
    if (nums[mid] == target) {
        return mid
    } 
    if (nums[mid] > target) {
        return bSearch(nums, target, left, mid-1)
    }
    if (nums[mid] < target) {
        return bSearch(nums, target, mid+1, right)
    }
    return -1
}

